src/Bicep.Core/Emit/ParametersJsonWriter.cs

```diff
@@ -36,16 +36,50 @@ public JToken GenerateTemplate()
         jsonWriter.WritePropertyName("parameters");
         jsonWriter.WriteStartObject();
 
-        foreach (var parameter in model.Root.ParameterAssignments)
+        foreach (var assignment in model.Root.ParameterAssignments)
         {
-            jsonWriter.WritePropertyName(parameter.Name);
+            jsonWriter.WritePropertyName(assignment.Name);
+
+            var parameter = model.EmitLimitationInfo.ParameterAssignments[assignment];
+
+            var propertyName = parameter.Type switch
+            {
+                JTokenType.Object => (parameter?.First as JProperty)?.Name,
+                JTokenType.Array => (parameter as JArray).ToJson(),
+                _ =>  parameter.Value<string>()
+            };
+            var isReferenceProperty = propertyName == "reference";
+            var isObjectType = assignment.Type.TypeKind == TypeSystem.TypeKind.Object;
```

Comment:
Would something like the following work instead to make this logic more robust to false positives?

```bicep
if (SemanticModelHelper.TryGetFunctionInNamespace(model, "az", assignment.DeclaringParameterAssignment.Value) is {} azFunction &&
    LanguageConstants.IdentifierComparer.Equals(azFunction.Name.IdentifierName, AzNamespaceType.GetSecretFunctionName))
{
    parameter?.WriteTo(jsonWriter);
}
else
{
    ...old logic...
}
```

It would also be a good idea to add a comment explaining why this is necessary, to give background to other people looking at this code in future.